x/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,92 @@
+package upgrade_test
```

Comment:
I'd recommend having unit tests for all the functions as well as this integration test.

Reply:
Yup for sure. Will add all the tests once we're happy with the design